### PR TITLE
Create and assign CaDeT domains

### PR DESCRIPTION
**Pull request by murdo-moj**
* Added a new source which creates domains from the CaDeT manifest
* Added a transformer which builds upon the dbt ingestion and assigns those domains to CaDet models
    * The transformer has been adapted from the existing [PatternAddDatasetDomain](https://github.com/datahub-project/datahub/blob/648fd459eb6b981a8283084fde4942fff6c9d6aa/metadata-ingestion/src/datahub/ingestion/transformer/dataset_domain.py#L141) transformer written by DataHub.
    * A urn is formed for every dataset from the manifest which maps to a domain.
* Domains in CaDeT are not assigned to sources, only models, so out of 505 datasets currently ingested, only ~310 have domains (the rest are sources).
* Changed dependency management to poetry for alignment with our other projects
* Adjusted dbt ingestion workflows to use new source and transformer
    * The cadet ingestion source (dbt) can't create domains, so the creation of domains happens before dbt ingestion

### CODE REVIEW

File: ingestion/create_derived_table_domains_source/source.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    def _get_manifest(self, manifest_s3_uri: str) -> Dict:
+        try:
+            s3 = boto3.client("s3")
+            s3_parts = manifest_s3_uri.split("/")
+            bucket_name = s3_parts[2]
+            file_key = "/".join(s3_parts[3:])
+            response = s3.get_object(Bucket=bucket_name, Key=file_key)
+            content = response['Body'].read().decode('utf-8')
+            manifest = json.loads(content)
```
Comment (murdo-moj) [inline ingestion/create_derived_table_domains_source/source.py:58]:
```suggestion
            manifest = json.loads(content, strict=False)
```

File: ingestion/transformers/assign_cadet_domains.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            raise
+        return manifest
+
+    def _get_domain_mapping(self, manifest) -> PatternDatasetDomainSemanticsConfig:
+        """Map regex patterns for tables to domains"""
+        nodes = manifest.get("nodes")
+        domain_mappings = {}
+
+        for node in nodes:
+            if manifest["nodes"][node]["resource_type"] == "seed":
```
Comment (murdo-moj) [inline ingestion/transformers/assign_cadet_domains.py:106]: If sources don't need domains we can add "source" here as well. This will reduce the size of the mapping dict so speed up ingestion

File: ingestion/cadet.yaml [hunk 3]
```diff
@@ -6,9 +6,7 @@ source:
     manifest_path: "s3://mojap-derived-tables/prod/run_artefacts/latest/target/manifest.json"
     catalog_path: "s3://mojap-derived-tables/prod/run_artefacts/latest/target/catalog.json"
     test_results_path: "s3://mojap-derived-tables/prod/run_artefacts/latest/target/run_results.json"
-    platform_instance: cadet
     target_platform: athena
-    target_platform_instance: athena_cadet
```
Comment (LavMatt) [inline ingestion/cadet.yaml:11]: I think we should retain the platform instances. Should we have different dbt sources in future it'll be useful to distinguish them.

**Review by MatMoore — COMMENTED** [review]
Haven't finished reviewing yet, but have left some comments inline

File: ingestion/create_derived_table_domains_source/source.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        for domain_name in self._get_domains(manifest):
+            mcp = self._make_domain(domain_name)
+            wu = MetadataWorkUnit("single_mcp", mcp=mcp)
+            self.report.report_workunit(wu)
+
+            yield wu
+
+    def _get_domains(self, manifest) -> set[str]:
+        return set(
+            manifest["nodes"][node]["fqn"][1]
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply ingestion/create_derived_table_domains_source/source.py:44]: That would be more readable and I considered it, but I left it so that no alterations were made to the CaDeT definitions. Another option is to transform the domain name on the front end side.

File: ingestion/transformers/assign_cadet_domains.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                continue
+            domain = manifest["nodes"][node].get("fqn", [])[1]
+            node_table_name = manifest["nodes"][node].get("fqn", [])[-1]
+            node_table_name_no_double_underscore = node_table_name.replace("__", ".")
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply ingestion/transformers/assign_cadet_domains.py:106]: It's not mandated by DBT that table names are structured like this, but it is [mandated by CaDeT](https://user-guidance.analytical-platform.service.justice.gov.uk/tools/create-a-derived-table/project-structure/#standard-directory-structure-and-naming-conventions).

> Models (.sql files) must be named by the database and table they relate to separated by double underscores, i.e., <database_name>__<table_name>.sql. This is because all models in the models directory must have a unique name.

[the remark this answers is not in the document]
Comment (murdo-moj) [reply ingestion/transformers/assign_cadet_domains.py:106]: The sources don't follow the same convention for domain or naming, so we'd need some alternate parsing logic for them if we wanted to assign a domain to them.

**Review by MatMoore — DISMISSED** [review]
Overall the approach looks good to me, nice work 👍🏻 

There's a lot going on here, so anything we can do to make it easier to maintain and debug would be great. But the new github workflow makes sense, and the tests look good, so think it's good to go once the other comments are resolved.

File: ingestion/create_derived_table_domains_source/source.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatMoore) [reply ingestion/create_derived_table_domains_source/source.py:44]: Yep, we can do that 👍🏻

File: tests/test_cadet_transformer.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        )
+
+        assert len(output) == 2
+        assert output[0] is not None
+        assert output[0].record is not None
+        assert isinstance(output[0].record, MetadataChangeProposalWrapper)
+        assert output[0].record.aspect is not None
+        assert isinstance(output[0].record.aspect, models.DomainsClass)
+        transformed_aspect = cast(list, output[0].record.aspect)
+        assert len(transformed_aspect.domains) == 1
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply tests/test_cadet_transformer.py:101]: I repurposed all of the tests from DataHub's tests... It's interesting to me that they've chosen to do all of these type checks and length checks rather than some more direct comparison. I'm unsure as to what advantage this has.

File: ingestion/transformers/assign_cadet_domains.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (murdo-moj) [reply ingestion/transformers/assign_cadet_domains.py:106]: I've extracted them to a method and added a comment 👍

File: tests/test_cadet_transformer.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        assert prison_domain in transformed_aspect.domains
+
+    # A bug in datahub's transformers is that domains don't overwrite correctly.
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply tests/test_cadet_transformer.py:104]: The test for this is super indirect and hard to follow:
* L15 `make_generic_dataset_mcp()` is called to make a blank dataset with a urn which matches an entry in `manifest.json`
* L114 Assigns the probation domain to this blannk dataset, before the transformer is applied to the dataset creation event
* The transformer then looks in `data/manifest.json` using boto3 and maps the urn to the prison domain and overwrites the probation domain for that dataset
* So the check at the end on L131/L132 checks the overwrite has happened

File: ingestion/transformers/assign_cadet_domains.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import json
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply ingestion/transformers/assign_cadet_domains.py:1]: It's not super well documented by DataHub but my understanding is that sources create entities, while transformers alter entity creation requests to add more aspects. So conceptually it fits better to alter the creation of all the datasets from dbt with another domain association upon the creation of the entity. You get the entities for free as part of the dbt ingestion, otherwise you'd have to do a lookup in the domain creation source.

It could probably work all in the source file, and the manifest wouldn't have to be read twice. Maybe I can pull out the code to read from s3 into a module?

[the remark this answers is not in the document]
Comment (MatMoore) [reply ingestion/transformers/assign_cadet_domains.py:1]: Yep I think that makes sense, and extracting the common code would be a good improvement 👍🏻

File: tests/test_cadet_transformer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MatMoore) [reply tests/test_cadet_transformer.py:101]: Either should work 🤷🏻‍♂️

File: tests/test_cadet_transformer.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        entityType=Urn.create_from_string(entity_urn).get_type(),
+        aspectName=aspect_name,
+        changeType="UPSERT",
+        aspect=aspect,
+    )
+
+
+def run_dataset_transformer_pipeline(
+    transformer_type: Type[DatasetTransformer],
+    aspect: Optional[builder.Aspect],
```
[the remark this answers is not in the document]
Comment (murdo-moj) [reply tests/test_cadet_transformer.py:31]: This is a copy of DataHub's implementation, which has a load of other transformers which do use mces as opposed to `MetadataChangeProposalWrapper` like our transformer uses. All the tests use these same helpers so having the flag there makes it clear if am mce or mcp is being used.

**Review by MatMoore — APPROVED** [review]
